### PR TITLE
Do not put noalias annotations by default

### LINKED ISSUE #54462: Incorrect code generation for nalgebra's Matrix::swap_rows()
The nalgebra linear algebra library has a swap_rows method which allows the user to swap two rows of a matrix. Unfortunately, I'm currently investigating [a code generation heisenbug which causes this method to corrupt the matrix data in some circumstances](https://github.com/rustsim/nalgebra/issues/396).

Given the UB-like symptoms, and the fact that the implementation of swap_rows takes multiple non-overlapping &mut references to the target matrix, I [wondered if this could be a violation of Rust's aliasing rules](https://github.com/nikomatsakis/rust-memory-model/issues/50). However, @nagisa confirmed that this is not not the case, and that the compiler is probably the culprit here. He identified the recent LLVM upgrade from LLVM 7 to LLVM 8 as a likely cause of this issue.

Here is a minimal reproducer of my problem:

```rust
extern crate nalgebra;

use nalgebra::Matrix3x4;

fn swappy() -> Matrix3x4<f32> {
    let mut mat = Matrix3x4::new(1., 2.,  3.,  4.,
                                 5., 6.,  7.,  8.,
                                 9., 10., 11., 12.);

    // NOTE: This printf makes the bug go away, suggesting UB or a codegen issue
    // println!("Result: {}", mat);

    for i in 0..2 {
        for j in i+1..3 {
            if mat[(j, 3)] > mat[(i, 3)] { mat.swap_rows(i, j); }
        }
    }

    mat
}

fn main() {
    let mat = swappy();
    println!("Result: {}", mat);
}
```

To reproduce the issue, you must build in release mode. The issue is also sensitive to the amount of codegen units in flight, therefore I strongly recommend building with codegen-units=1 as well.

I expect the following output:

```
  ┌             ┐
  │  9 10 11 12 │
  │  5  6  7  8 │
  │  1  2  3  4 │
  └             ┘
```

Instead, on my systems (nalgebra 0.16.2, rust 1.29, Ivy Bridge & Haswell CPUs) I get the following output:

```
  ┌             ┐
  │  9 10 11 12 │
  │  5  6  7  8 │
  │  1  6  7  4 │
  └             ┘
```

### PR DESCRIPTION
**Pull request by nagisa**
This will be re-enabled sooner or later depending on results of further
investigation.

Fixes #54462 

Beta backport is: #54640

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_codegen_llvm/type_of.rs [hunk 1]
```diff
@@ -433,12 +431,19 @@ impl<'tcx> LayoutLlvmExt<'tcx> for TyLayout<'tcx> {
...
                     hir::MutMutable => {
-                        // Only emit noalias annotations for LLVM >= 6 or in panic=abort
-                        // mode, as prior versions had many bugs in conjunction with
-                        // unwinding. See also issue #31681.
+                        // Previously we would only emit noalias annotations for LLVM >= 6 or in
+                        // panic=abort mode. That was deemed right, as prior versions had many bugs
+                        // in conjunction with unwinding, but later versions didn’t seem to have
+                        // said issues. See issue #31681.
+                        //
+                        // Alas, later on we encountered a case where nomangle would generate wrong
```
Comment (uberjay) [inline src/librustc_codegen_llvm/type_of.rs]: nomangle -> noalias?

Comment (nagisa) [reply src/librustc_codegen_llvm/type_of.rs]: Ew, that’s embarrassing.

**Comment by kennytm** [comment]
@bors r-

The codegen test `codegen\function-arguments.rs` failed on x86_64-pc-windows-msvc. Error log: https://github.com/rust-lang/rust/pull/54660#issuecomment-425623413

```
C:\projects\rust\src/test\codegen\function-arguments.rs:56:11: error: CHECK: expected string not found in input
// CHECK: @mutable_unsafe_borrow(i16* noalias dereferenceable(2) %arg0)
          ^
C:\projects\rust\build\x86_64-pc-windows-msvc\test\codegen\function-arguments\function-arguments.ll:272:59: note: scanning from here
define void @unsafe_borrow(i16* dereferenceable(2) %arg0) unnamed_addr #0 {
                                                          ^
C:\projects\rust\build\x86_64-pc-windows-msvc\test\codegen\function-arguments\function-arguments.ll:278:5: note: possible intended match here
define void @mutable_unsafe_borrow(i16* dereferenceable(2) %arg0) unnamed_addr #0 {
    ^
```

**Comment by nagisa** [comment]
Beta backport for this has landed as well, so just marking as beta-accepted.

**Comment by pnkfelix** [comment]
We discussed this (for a perhaps surprisingly long time) at the T-compiler meeting.

No one objects to the idea of backporting this PR to stable in terms of the benefit of this change and the risk involved.

There was some debate about whether a point release is *justifiable* for this particular bug.

Namely, it was suggested that point releases should be reserved for:
  * security fixes
  * accidental stabilizations
  * fixes to *new* (and "severe") regressions.

But there was enough debate about whether that *is* the policy to then lead us to say that we should resolve that question separately from the approval process for this PR.

So, tagging as stable-accepted.

**Pull request merged**